File: content/snippets/cpp/design_patterns/abstract_factory.cpp
#include <iostream>
#include <memory>
#include <string>

struct Button {
    virtual ~Button() = default;
    virtual std::string render() const = 0;
};

struct Dialog {
    virtual ~Dialog() = default;
    virtual std::string render() const = 0;
};

struct WidgetFactory {
    virtual ~WidgetFactory() = default;
    virtual std::unique_ptr<Button> createButton() const = 0;
    virtual std::unique_ptr<Dialog> createDialog() const = 0;
};

struct LightButton : Button {
    std::string render() const override { return "light button"; }
};

struct LightDialog : Dialog {
    std::string render() const override { return "light dialog"; }
};

struct DarkButton : Button {
    std::string render() const override { return "dark button"; }
};

struct DarkDialog : Dialog {
    std::string render() const override { return "dark dialog"; }
};

struct LightFactory : WidgetFactory {
    std::unique_ptr<Button> createButton() const override { return std::make_unique<LightButton>(); }
    std::unique_ptr<Dialog> createDialog() const override { return std::make_unique<LightDialog>(); }
};

struct DarkFactory : WidgetFactory {
    std::unique_ptr<Button> createButton() const override { return std::make_unique<DarkButton>(); }
    std::unique_ptr<Dialog> createDialog() const override { return std::make_unique<DarkDialog>(); }
};

void showScreen(const WidgetFactory& factory) {
    auto button = factory.createButton();
    auto dialog = factory.createDialog();
    std::cout << button->render() << " + " << dialog->render() << '\n';
}

int main() {
    LightFactory light;
    DarkFactory dark;
    showScreen(light);
    showScreen(dark);
}

File: content/snippets/scala2/design_patterns/abstract_factory.scala
// Abstract Factory
trait UIElement

case class Button(label: String) extends UIElement
case class Checkbox(label: String) extends UIElement

trait UIFactory {
  def createButton(label: String): Button
  def createCheckbox(label: String): Checkbox
}

object LightThemeFactory extends UIFactory {
  def createButton(label: String): Button = Button(s"Light Button: $label")
  def createCheckbox(label: String): Checkbox = Checkbox(s"Light Checkbox: $label")
}

object DarkThemeFactory extends UIFactory {
  def createButton(label: String): Button = Button(s"Dark Button: $label")
  def createCheckbox(label: String): Checkbox = Checkbox(s"Dark Checkbox: $label")
}

object Main extends App {
  val lightFactory: UIFactory = LightThemeFactory
  val darkFactory: UIFactory = DarkThemeFactory
  
  println(lightFactory.createButton("OK"))
  println(darkFactory.createButton("OK"))
}
